File: test-parser.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');

// Exact copy of parseGenerationTable from scraper with debugging
function parseGenerationTable(html) {
    const periods = [];

    // Target GridView1 specifically
    const gridMatch = html.match(/<table[^>]+id="GridView1"[^>]*>([\s\S]*?)<\/table>/i);
    if (!gridMatch) {
        console.log('  ⚠️ GridView1 table not found in HTML');
        return periods;
    }

    console.log('  ✅ GridView1 table found');
    const tableHtml = gridMatch[1];
    console.log(`  Table HTML length: ${tableHtml.length} chars`);

    // Remove the caption section entirely to avoid nested table rows
    const tableWithoutCaption = tableHtml.replace(/<caption[^>]*>[\s\S]*?<\/caption>/gi, '');
    console.log(`  After caption removal: ${tableWithoutCaption.length} chars`);

    const rows = [...tableWithoutCaption.matchAll(/<tr[^>]*>([\s\S]*?)<\/tr>/gi)];
    console.log(`  ✅ Found ${rows.length} rows (including header)`);

    // Show first few rows to debug
    for (let i = 0; i < Math.min(3, rows.length); i++) {
        const rowHtml = rows[i][0].substring(0, 100);
        console.log(`  Row ${i} preview: ${rowHtml}...`);
    }

    // Skip the header row (index 0)
    for (let i = 1; i < rows.length; i++) {
        const cells = [...rows[i][1].matchAll(/<td[^>]*>([\s\S]*?)<\/td>/gi)]
            .map(c => c[1].replace(/<[^>]+>/g, '').trim());

        if (i <= 5) {
            console.log(`  Row ${i}: Found ${cells.length} cells - [${cells.join(', ')}]`);
        }

        if (cells.length >= 2) {
            const timeText = cells[0];
            const genMatch = cells[1].match(/^[\d.]+$/);

            if (i <= 5) {
                console.log(`    timeText="${timeText}", genText="${cells[1]}", genMatch=${!!genMatch}`);
            }

            if (timeText && genMatch) {
                const generation = parseFloat(cells[1]);
                periods.push({
                    time: timeText,
                    generation,
                    status: generation > 50 ? 'peak' : generation > 10 ? 'active' : 'base',
                    source: 'USACE Real-time'
                });
            }
        }
    }

    console.log(`  ✅ Parsed ${periods.length} periods total`);
    return periods;
}

// Read the debug HTML file
const debugFile = path.join(__dirname, 'data', 'debug-date-response.html');

if (!fs.existsSync(debugFile)) {
    console.error('❌ Debug file not found:', debugFile);
    console.log('Run the scraper first to generate debug-date-response.html');
    process.exit(1);
}

const html = fs.readFileSync(debugFile, 'utf8');

console.log('='.repeat(60));
console.log('Testing parseGenerationTable with debug HTML');
console.log('='.repeat(60));
console.log();

const periods = parseGenerationTable(html);

console.log();
console.log('='.repeat(60));
console.log(`📊 RESULT: Found ${periods.length} periods`);
console.log('='.repeat(60));

if (periods.length > 0) {
    console.log('\nFirst 5 periods:');
    periods.slice(0, 5).forEach((p, i) => {
        console.log(`  ${i + 1}. ${p.time}: ${p.generation} MW (${p.status})`);
    });
    console.log();
    console.log('✅ SUCCESS - Parser is working correctly!');
} else {
    console.log();
    console.log('❌ FAILURE - No periods were parsed');
    console.log('Check the debug output above to see what went wrong');
}

File: test-parser.py
#!/usr/bin/env python3

import re
import json

# Read the debug HTML
with open('data/debug-date-response.html', 'r') as f:
    html = f.read()

print('='*60)
print('Testing HTML Parser')
print('='*60)

# Step 1: Find GridView1 table (use greedy match to get to the real closing tag, not the nested one)
grid_match = re.search(r'<table[^>]+id="GridView1"[^>]*>([\s\S]*)</table>', html, re.IGNORECASE)

if not grid_match:
    print('❌ GridView1 table not found!')
    exit(1)

print('✅ GridView1 table found')
table_html = grid_match.group(1)
print(f'   Table HTML length: {len(table_html)} chars')

# Step 2: Remove caption
table_without_caption = re.sub(r'<caption[^>]*>[\s\S]*?</caption>', '', table_html, flags=re.IGNORECASE)
print(f'✅ After caption removal: {len(table_without_caption)} chars')

# Step 3: Find all rows
rows = re.findall(r'<tr[^>]*>([\s\S]*?)</tr>', table_without_caption, re.IGNORECASE)
print(f'✅ Found {len(rows)} rows (including header)')

# Step 4: Parse data rows
periods = []

for i, row_content in enumerate(rows):
    if i == 0:  # Skip header row
        continue

    # Find all td cells
    cells = re.findall(r'<td[^>]*>([\s\S]*?)</td>', row_content, re.IGNORECASE)
    # Remove any HTML tags and trim
    cells = [re.sub(r'<[^>]+>', '', cell).strip() for cell in cells]

    if i <= 3:
        print(f'   Row {i}: Found {len(cells)} cells - {cells}')

    if len(cells) >= 2:
        time_text = cells[0]
        gen_text = cells[1]

        # Check if generation matches digit pattern
        gen_match = re.match(r'^[\d.]+$', gen_text)

        if i <= 3:
            print(f'      timeText="{time_text}", genText="{gen_text}", genMatch={bool(gen_match)}')

        if time_text and gen_match:
            generation = float(gen_text)
            periods.append({
                'time': time_text,
                'generation': generation,
                'status': 'peak' if generation > 50 else ('active' if generation > 10 else 'base'),
                'source': 'USACE Real-time'
            })

print()
print('='*60)
print(f'📊 RESULT: Found {len(periods)} periods')
print('='*60)

if periods:
    print('\nFirst 5 periods:')
    for i, p in enumerate(periods[:5]):
        print(f'   {i+1}. {p["time"]}: {p["generation"]} MW ({p["status"]})')
    print('\n✅ SUCCESS - Parser logic is correct!')
else:
    print('\n❌ FAILURE - No periods were parsed')
    print('This means there is a bug in the JavaScript parsing code')
